File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], sub: Optional[TreeNode]) -> bool:
        if not sub:
            return True
        
        if not root:
            return False
        
        same = self.isSame(root, sub)
        if same:
            return True
        
        left = self.isSubtree(root.left, sub)
        right = self.isSubtree(root.right, sub)

        return left or right


    def isSame(self, root, sub):
        if not root and not sub:
            return True 
        
        if root and sub and root.val == sub.val:
            left = self.isSame(root.left, sub.left)
            right = self.isSame(root.right, sub.right)

            return left and right
        
        return False

File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.swift
class Solution {
    func isValidBST(_ root: TreeNode?) -> Bool {
        func isValid(
            _ node: TreeNode?,
            _ minVal: Int?,
            _ maxVal: Int?
        ) -> Bool {
            guard let node else { return true }

            if let minVal, node.val <= minVal {
                return false
            }

            if let maxVal, node.val >= maxVal {
                return false
            }

            return isValid(node.left, minVal, node.val)
                && isValid(node.right, node.val, maxVal)
        }

        return isValid(root, nil, nil)
    }
}